### ISSUE TITLE
Align on an output format for detected dependencies

### ISSUE THREAD

**Issue by sschuberth** [microsoft/component-detection#42]
I realized this tool is currently using its own JSON format for reporting dependencies, but I wonder whether instead it would make more sense to write out standardized formats like [SPDX](https://spdx.github.io/spdx-spec/) or [CycloneDX](https://cyclonedx.org/specification/overview/), or at least try to align with existing formats of similar tools like the [ORT Analyzer](https://github.com/oss-review-toolkit/ort#analyzer). (*Disclaimer*: I'm the founder of ORT.)

**Comment by brphelps** [comment]
Supporting multiple formats could make sense for something like this -- maximizing human usability in the default format doesn't seem like a _bad_ thing.

**Comment by sschuberth** [comment]
> maximizing human usability in the default format doesn't seem like a _bad_ thing.

Agreed. Though I'd argue there are better formats for human usability than JSON 😉

**Comment by coderpatros** [comment]
I did raise #22 for this. Although the title indicates CycloneDX output format, it can leverage the CycloneDX/SPDX interop library to also be able to support SPDX output.

**Comment by JamieMagee** [comment]
I've added this as a topic for our first community meeting. See #49.

**Comment by sschuberth** [comment]
> another team at Microsoft is building an SBOM generation tool, taking a dependency on Component Detection to get dependency info. They plan to open source it "soon".

Ah, I read about that tool [over here](https://devblogs.microsoft.com/engineering-at-microsoft/generating-software-bills-of-materials-sboms-with-spdx-at-microsoft/), and I actually though *this* project is the tool that Microsoft announced over there. Thanks for the clarification that it's yet another tool!

> Duplication of work wouldn't be great, but maximising the utility of Component Detection for other users outside of Microsoft would be great!

Speaking of duplication of work... the [ORT](https://github.com/oss-review-toolkit/ort#readme) community certainly would have appreciated contributions to its [analyzer tool](https://github.com/oss-review-toolkit/ort#analyzer) (esp. regarding NuGet support, as we're not .NET people) instead of creating yet another tool with a very similar purpose :wink:

**Comment by JamieMagee** [comment]
We discussed this topic in our [community meeting on Wednesday 26th January](https://github.com/microsoft/component-detection/issues/49) and we've decided to go ahead with adding support for SPDX and CycloneDX support. I don't know _when_ exactly, but it's on our backlog now.

**Comment by coderpatros** [comment]
That's great news @JamieMagee 

I've poked around in the code base a bit. I think an appropriate approach would be adding a "ManifestFileFormat" option. Default would be the existing format, with options for CycloneDX and SPDX JSON formats.

Then add handling for it in the BcdeScanCommandService.WriteComponentManifest(detectionArguments, scanResult) method.

The CycloneDX project already has a library that will convert from CycloneDX to SPDX format. So the only extra bit would be adding a method to convert the internal format to CycloneDX. Perhaps a CDXFormatService?

Does that sound like a reasonable approach to you?
